File: research/cryptanalysis/transcript/typescript/transcript.ts
import { keccak_256 } from "../../../../packages/sdk/node_modules/@noble/hashes/sha3.js";

export const MODULUS = 2_013_265_921;
export type Candidate = "T0" | "T1" | "T2" | "T3";
export type ItemType = 1 | 2 | 3;
export interface Metrics { keccakCalls: number; hashedBytes: number; copiedBytes: number; peakFrameBytes: number }
export interface Snapshot { label: string; operation: "absorb" | "squeeze"; state: string; pendingItems: number; pendingBytes: number; outputBytes: number; squeezeCounter: string; rejectedWords: number }

const encoder = new TextEncoder();
const VERSION: Record<Candidate, Uint8Array> = {
  T0: encoder.encode("PQTCT0-01"),
  T1: encoder.encode("PQTCT1-01"),
  T2: encoder.encode("PQTCT2-01"),
  T3: encoder.encode("PQTCT3-01"),
};

function concat(...parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let at = 0;
  for (const part of parts) { out.set(part, at); at += part.length; }
  return out;
}
export function u16(n: number): Uint8Array {
  if (!Number.isSafeInteger(n) || n < 0 || n > 0xffff) throw new Error("u16 out of range");
  return Uint8Array.of(n >>> 8, n);
}
export function u32(n: number): Uint8Array {
  if (!Number.isSafeInteger(n) || n < 0 || n > 0xffff_ffff) throw new Error("u32 out of range");
  return Uint8Array.of(n >>> 24, n >>> 16, n >>> 8, n);
}
export function u64(n: bigint): Uint8Array {
  if (n < 0n || n > 0xffff_ffff_ffff_ffffn) throw new Error("u64 out of range");
  const out = new Uint8Array(8);
  for (let i = 7; i >= 0; i--) { out[i] = Number(n & 255n); n >>= 8n; }
  return out;
}
export function field(n: number): Uint8Array {
  if (!Number.isSafeInteger(n) || n < 0 || n >= MODULUS) throw new Error(`noncanonical field ${n}`);
  return u32(n);
}
export function hex(bytes: Uint8Array): string { return Buffer.from(bytes).toString("hex"); }
export function fromHex(value: string, bytes?: number): Uint8Array {
  if (!/^(?:[0-9a-f]{2})*$/i.test(value)) throw new Error("noncanonical hex");
  const out = Uint8Array.from(Buffer.from(value, "hex"));
  if (bytes !== undefined && out.length !== bytes) throw new Error(`expected ${bytes} bytes`);
  return out;
}
function be32(bytes: Uint8Array): number {
  return ((bytes[0]! * 0x1000000) + (bytes[1]! << 16) + (bytes[2]! << 8) + bytes[3]!) >>> 0;
}

export class Transcript {
  variant: Candidate;
  metrics: Metrics = { keccakCalls: 0, hashedBytes: 0, copiedBytes: 0, peakFrameBytes: 0 };
  events: Snapshot[] = [];
  private stateBytes: Uint8Array;
  private output = new Uint8Array();
  private outputAt = 0;
  private counter = 0n;
  private pending: Uint8Array[] = [];
  private pendingLength = 0;
  private rejected = 0;

  constructor(variant: Candidate, parameterDigest: Uint8Array, publicValues: number[], productionCompatibleT0 = false) {
    this.variant = variant;
    if (parameterDigest.length !== 64) throw new Error("parameter digest must be 64 bytes");
    publicValues.forEach(field);
    const fields = concat(...publicValues.map(field));
    const payload = productionCompatibleT0 && variant === "T0"
      ? concat(parameterDigest, fields)
      : concat(VERSION[variant], parameterDigest, u32(publicValues.length), fields);
    this.stateBytes = this.k512(0x42, payload);
    this.events.push(this.snapshot("init", "absorb"));
  }

  private k512(tag: number, payload: Uint8Array): Uint8Array {
    const leftInput = concat(Uint8Array.of(0, tag), payload);
    const rightInput = concat(Uint8Array.of(1, tag), payload);
    this.metrics.keccakCalls += 2;
    this.metrics.hashedBytes += leftInput.length + rightInput.length;
    this.metrics.copiedBytes += leftInput.length + rightInput.length + 64;
    return concat(keccak_256(leftInput), keccak_256(rightInput));
  }
  private resetOutput(): void { this.output = new Uint8Array(); this.outputAt = 0; this.counter = 0n; }
  private typed(type: ItemType, payload: Uint8Array, label: string): void {
    if (type === 1 && payload.length !== 4) throw new Error("scalar width");
    if (type === 2 && payload.length !== 64) throw new Error("commitment width");
    if (type === 3) decodeFieldArray(payload);
    if (this.variant === "T2") {
      const digest = this.k512(0x45, concat(VERSION.T2, Uint8Array.of(type), u32(payload.length), payload));
      this.stateBytes = this.k512(0x43, concat(this.stateBytes, digest));
      this.resetOutput();
    } else if (this.variant === "T3") {
      const item = concat(Uint8Array.of(type), u32(payload.length), payload);
      this.pending.push(item); this.pendingLength += item.length;
      this.metrics.copiedBytes += item.length;
      this.metrics.peakFrameBytes = Math.max(this.metrics.peakFrameBytes, this.pendingLength);
    } else {
      this.stateBytes = this.k512(0x43, concat(this.stateBytes, Uint8Array.of(type), u32(payload.length), payload));
      this.resetOutput();
    }
    this.events.push(this.snapshot(label, "absorb"));
  }
  absorbField(value: number, label: string): void { this.typed(1, field(value), label); }
  absorbCommitment(value: Uint8Array, label: string): void {
    if (value.length !== 64) throw new Error("commitment must be 64 bytes");
    this.typed(2, value, label);
  }
  absorbFields(values: number[], label: string): void {
    values.forEach(field);
    if (this.variant === "T0") {
      values.forEach((v, i) => this.absorbField(v, `${label}[${i}]`));
      return;
    }
    this.typed(3, concat(u32(values.length), ...values.map(field)), label);
  }
  private flush(): void {
    if (this.variant !== "T3" || this.pending.length === 0) return;
    const items = concat(...this.pending);
    const payload = concat(VERSION.T3, u32(this.pending.length), u32(items.length), items);
    this.stateBytes = this.k512(0x46, concat(this.stateBytes, payload));
    this.pending = []; this.pendingLength = 0; this.resetOutput();
  }
  private bytes(count: number): Uint8Array {
    this.flush();
    const out = new Uint8Array(count);
    for (let i = 0; i < count; i++) {
      if (this.outputAt === this.output.length) {
        this.output = this.k512(0x44, concat(this.stateBytes, u64(this.counter++)));
        this.outputAt = 0;
      }
      out[i] = this.output[this.outputAt++]!;
    }
    return out;
  }
  sampleField(label: string): number {
    for (;;) {
      const word = be32(this.bytes(4)) & 0x7fffffff;
      if (word < MODULUS) { this.events.push(this.snapshot(label, "squeeze")); return word; }
      this.rejected++;
    }
  }
  sampleExtension(label: string): number[] {
    const value = [0, 1, 2, 3].map((i) => this.sampleField(`${label}.c${i}`));
    return value;
  }
  sampleBits(bits: number, label: string): number {
    if (!Number.isInteger(bits) || bits < 0 || bits > 31) throw new Error("bits out of range");
    const word = be32(this.bytes(4));
    const value = bits === 0 ? 0 : word & (2 ** bits - 1);
    this.events.push(this.snapshot(label, "squeeze"));
    return value;
  }
  clone(): Transcript {
    const out = Object.create(Transcript.prototype) as Transcript;
    Object.assign(out, this);
    out.metrics = { ...this.metrics };
    out.events = [...this.events];
    out.stateBytes = this.stateBytes.slice(); out.output = this.output.slice();
    out.pending = this.pending.map((p) => p.slice());
    return out;
  }
  adopt(other: Transcript): void {
    this.stateBytes = other.stateBytes; this.output = other.output; this.outputAt = other.outputAt;
    this.counter = other.counter; this.pending = other.pending; this.pendingLength = other.pendingLength;
    this.rejected = other.rejected; Object.assign(this.metrics, other.metrics);
    this.events.splice(0, this.events.length, ...other.events);
  }
  checkWitness(bits: number, witness: number, label: string): boolean {
    if (bits === 0) return true;
    this.absorbField(witness, `${label}.witness`);
    return this.sampleBits(bits, `${label}.check`) === 0;
  }
  grind(bits: number, label: string): number {
    if (bits === 0) return 0;
    for (let witness = 0; witness < MODULUS; witness++) {
      const trial = this.clone();
      trial.absorbField(witness, `${label}.witness`);
      if (trial.sampleBits(bits, `${label}.check`) === 0) { this.adopt(trial); return witness; }
    }
    throw new Error("no PoW witness");
  }
  state(): Uint8Array { this.flush(); return this.stateBytes.slice(); }
  snapshot(label: string, operation: "absorb" | "squeeze"): Snapshot {
    return { label, operation, state: hex(this.stateBytes), pendingItems: this.pending.length, pendingBytes: this.pendingLength,
      outputBytes: this.output.length - this.outputAt, squeezeCounter: this.counter.toString(), rejectedWords: this.rejected };
  }
}

export function encodeFieldArray(values: number[]): Uint8Array { values.forEach(field); return concat(u32(values.length), ...values.map(field)); }
export function decodeFieldArray(payload: Uint8Array): number[] {
  if (payload.length < 4) throw new Error("truncated count");
  const count = be32(payload.subarray(0, 4));
  if (payload.length !== 4 + count * 4) throw new Error("array count/length mismatch or trailing bytes");
  const out: number[] = [];
  for (let at = 4; at < payload.length; at += 4) {
    const value = be32(payload.subarray(at, at + 4)); field(value); out.push(value);
  }
  return out;
}

function abiWordAscii(value: string): Uint8Array {
  const bytes = new TextEncoder().encode(value);
  if (bytes.length > 32) throw new Error("ABI word overflow");
  return concat(bytes, new Uint8Array(32 - bytes.length));
}
function abiWordField(value: number): Uint8Array { return concat(new Uint8Array(28), field(value)); }
export function statementKey(parameter: Uint8Array, publicValues: number[]): Uint8Array {
  if (parameter.length !== 64 || publicValues.length !== 64) throw new Error("full-width statement required");
  publicValues.forEach(field);
  return keccak_256(concat(abiWordAscii("PQTC.V3.STATEMENT"), parameter, ...publicValues.map(abiWordField)));
}
export function checkpointDigest(key: Uint8Array, payload: Uint8Array): Uint8Array {
  if (key.length !== 32) throw new Error("key width");
  return keccak_256(concat(abiWordAscii("PQTC.V3.CHECKPOINT"), key, keccak_256(payload)));
}
export function proofId(key: Uint8Array, partA: Uint8Array): Uint8Array {
  if (key.length !== 32) throw new Error("key width");
  return keccak_256(concat(abiWordAscii("PQTC.V3.PROOF"), key, keccak_256(partA)));
}
export interface FullDigestRecord { parameter: Uint8Array; statementDigests: Uint8Array[]; globalDigest: Uint8Array; coreProofDigest: Uint8Array }
function requireStatementDigests(publicValues: number[], digests: Uint8Array[]): void {
  if (digests.length !== 4 || digests.some((digest) => digest.length !== 64)) throw new Error("four full statement digests required");
  const encodedFields = digests.flatMap((digest) => Array.from({ length: 16 }, (_, i) => be32(digest.subarray(i * 4, i * 4 + 4))));
  if (encodedFields.length !== publicValues.length || encodedFields.some((value, i) => value !== publicValues[i])) throw new Error("statement digest/public field mismatch");
}
export class FullDigestStore {
  private records = new Map<string, FullDigestRecord>();
  put(publicValues: number[], record: FullDigestRecord): Uint8Array {
    if (record.parameter.length !== 64 || record.statementDigests.length !== 4 || record.statementDigests.some((d) => d.length !== 64)
      || record.globalDigest.length !== 32 || record.coreProofDigest.length !== 32) throw new Error("full-width record required");
    requireStatementDigests(publicValues, record.statementDigests);
    const key = statementKey(record.parameter, publicValues);
    this.records.set(hex(key), { parameter: record.parameter.slice(), statementDigests: record.statementDigests.map((d) => d.slice()), globalDigest: record.globalDigest.slice(), coreProofDigest: record.coreProofDigest.slice() });
    return key;
  }
  getChecked(key: Uint8Array, publicValues: number[], supplied: FullDigestRecord): FullDigestRecord {
    const stored = this.records.get(hex(key)); if (!stored) throw new Error("unknown lookup key");
    if (hex(statementKey(stored.parameter, publicValues)) !== hex(key)) throw new Error("stored full statement/key mismatch");
    requireStatementDigests(publicValues, stored.statementDigests);
    const all = (r: FullDigestRecord) => [r.parameter, ...r.statementDigests, r.globalDigest, r.coreProofDigest].map(hex);
    if (JSON.stringify(all(stored)) !== JSON.stringify(all(supplied))) throw new Error("full-width digest mismatch");
    return stored;
  }
}

File: research/cryptanalysis/transcript/rust/transcript.rs
use sha3::{Digest, Keccak256};
use std::collections::{HashMap, VecDeque};

pub const BABY_BEAR_MODULUS: u32 = 2_013_265_921;
const INIT: u8 = 0x42;
const ABSORB: u8 = 0x43;
const SQUEEZE: u8 = 0x44;
const ITEM_DIGEST: u8 = 0x45;
const BOUNDARY_FRAME: u8 = 0x46;
const FIELD: u8 = 1;
const COMMITMENT: u8 = 2;
const FIELD_ARRAY: u8 = 3;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Variant { T0, T1, T2, T3 }
impl Variant {
    pub const fn version(self) -> &'static [u8; 9] {
        match self { Self::T0 => b"PQTCT0-01", Self::T1 => b"PQTCT1-01", Self::T2 => b"PQTCT2-01", Self::T3 => b"PQTCT3-01" }
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Metrics { pub keccak_calls: u64, pub hashed_bytes: u64, pub copied_bytes: u64, pub peak_frame_bytes: usize }

#[derive(Clone, Debug)]
pub struct Transcript {
    variant: Variant,
    state: [u8; 64],
    output: VecDeque<u8>,
    squeeze_counter: u64,
    pending: Vec<Vec<u8>>,
    pending_bytes: usize,
    pub rejected_words: u64,
    pub metrics: Metrics,
}

impl Transcript {
    pub fn new(variant: Variant, parameter: &[u8; 64], public_values: &[u32], production_compatible_t0: bool) -> Result<Self, &'static str> {
        for &value in public_values { canonical(value)?; }
        let mut payload = Vec::with_capacity(9 + 64 + 4 + public_values.len() * 4);
        if !(production_compatible_t0 && variant == Variant::T0) {
            payload.extend_from_slice(variant.version());
        }
        payload.extend_from_slice(parameter);
        if !(production_compatible_t0 && variant == Variant::T0) {
            payload.extend_from_slice(&u32::try_from(public_values.len()).map_err(|_| "public count overflow")?.to_be_bytes());
        }
        for &value in public_values { payload.extend_from_slice(&value.to_be_bytes()); }
        let mut transcript = Self { variant, state: [0; 64], output: VecDeque::new(), squeeze_counter: 0, pending: Vec::new(), pending_bytes: 0, rejected_words: 0, metrics: Metrics::default() };
        transcript.state = transcript.k512(INIT, &payload);
        Ok(transcript)
    }

    fn k512(&mut self, tag: u8, payload: &[u8]) -> [u8; 64] {
        let mut left = Keccak256::new(); left.update([0, tag]); left.update(payload);
        let mut right = Keccak256::new(); right.update([1, tag]); right.update(payload);
        let mut out = [0u8; 64]; out[..32].copy_from_slice(&left.finalize()); out[32..].copy_from_slice(&right.finalize());
        self.metrics.keccak_calls += 2;
        self.metrics.hashed_bytes += 2 * (2 + payload.len()) as u64;
        self.metrics.copied_bytes += (2 * (2 + payload.len()) + 64) as u64;
        out
    }

    fn reset_output(&mut self) { self.output.clear(); self.squeeze_counter = 0; }

    fn typed(&mut self, item_type: u8, payload: &[u8]) -> Result<(), &'static str> {
        validate_item(item_type, payload)?;
        match self.variant {
            Variant::T2 => {
                let mut item = Vec::with_capacity(9 + 1 + 4 + payload.len());
                item.extend_from_slice(self.variant.version()); item.push(item_type);
                item.extend_from_slice(&u32::try_from(payload.len()).map_err(|_| "item too long")?.to_be_bytes()); item.extend_from_slice(payload);
                let digest = self.k512(ITEM_DIGEST, &item);
                let mut state_item = Vec::with_capacity(128); state_item.extend_from_slice(&self.state); state_item.extend_from_slice(&digest);
                self.state = self.k512(ABSORB, &state_item); self.reset_output();
            }
            Variant::T3 => {
                let mut item = Vec::with_capacity(5 + payload.len()); item.push(item_type);
                item.extend_from_slice(&u32::try_from(payload.len()).map_err(|_| "item too long")?.to_be_bytes()); item.extend_from_slice(payload);
                self.pending_bytes += item.len(); self.metrics.copied_bytes += item.len() as u64;
                self.metrics.peak_frame_bytes = self.metrics.peak_frame_bytes.max(self.pending_bytes); self.pending.push(item);
            }
            Variant::T0 | Variant::T1 => {
                let mut input = Vec::with_capacity(69 + payload.len()); input.extend_from_slice(&self.state); input.push(item_type);
                input.extend_from_slice(&u32::try_from(payload.len()).map_err(|_| "item too long")?.to_be_bytes()); input.extend_from_slice(payload);
                self.state = self.k512(ABSORB, &input); self.reset_output();
            }
        }
        Ok(())
    }

    pub fn absorb_field(&mut self, value: u32) -> Result<(), &'static str> { canonical(value)?; self.typed(FIELD, &value.to_be_bytes()) }
    pub fn absorb_commitment(&mut self, digest: &[u8; 64]) -> Result<(), &'static str> { self.typed(COMMITMENT, digest) }
    pub fn absorb_fields(&mut self, values: &[u32]) -> Result<(), &'static str> {
        for &value in values { canonical(value)?; }
        if self.variant == Variant::T0 { for &value in values { self.absorb_field(value)?; } return Ok(()); }
        let mut payload = Vec::with_capacity(4 + values.len() * 4);
        payload.extend_from_slice(&u32::try_from(values.len()).map_err(|_| "array count overflow")?.to_be_bytes());
        for &value in values { payload.extend_from_slice(&value.to_be_bytes()); }
        self.typed(FIELD_ARRAY, &payload)
    }

    fn flush(&mut self) -> Result<(), &'static str> {
        if self.variant != Variant::T3 || self.pending.is_empty() { return Ok(()); }
        let mut items = Vec::with_capacity(self.pending_bytes); for item in &self.pending { items.extend_from_slice(item); }
        let mut frame = Vec::with_capacity(17 + items.len()); frame.extend_from_slice(self.variant.version());
        frame.extend_from_slice(&u32::try_from(self.pending.len()).map_err(|_| "frame count overflow")?.to_be_bytes());
        frame.extend_from_slice(&u32::try_from(items.len()).map_err(|_| "frame length overflow")?.to_be_bytes()); frame.extend_from_slice(&items);
        let mut input = Vec::with_capacity(64 + frame.len()); input.extend_from_slice(&self.state); input.extend_from_slice(&frame);
        self.state = self.k512(BOUNDARY_FRAME, &input); self.pending.clear(); self.pending_bytes = 0; self.reset_output(); Ok(())
    }

    fn take_bytes<const N: usize>(&mut self) -> Result<[u8; N], &'static str> {
        self.flush()?; let mut out = [0u8; N];
        for byte in &mut out {
            if self.output.is_empty() {
                let mut payload = [0u8; 72]; payload[..64].copy_from_slice(&self.state); payload[64..].copy_from_slice(&self.squeeze_counter.to_be_bytes());
                self.squeeze_counter = self.squeeze_counter.checked_add(1).ok_or("squeeze counter overflow")?;
                let block = self.k512(SQUEEZE, &payload);
                self.output.extend(block);
            }
            *byte = self.output.pop_front().ok_or("empty squeeze")?;
        }
        Ok(out)
    }

    pub fn sample_field(&mut self) -> Result<u32, &'static str> {
        loop { let value = u32::from_be_bytes(self.take_bytes()?) & 0x7fff_ffff; if value < BABY_BEAR_MODULUS { return Ok(value); } self.rejected_words += 1; }
    }
    pub fn sample_extension(&mut self) -> Result<[u32; 4], &'static str> {
        Ok([self.sample_field()?, self.sample_field()?, self.sample_field()?, self.sample_field()?])
    }
    pub fn sample_bits(&mut self, bits: usize) -> Result<u32, &'static str> {
        if bits > 31 { return Err("bits out of range"); }
        let value = u32::from_be_bytes(self.take_bytes()?); Ok(if bits == 0 { 0 } else { value & ((1u32 << bits) - 1) })
    }
    pub fn check_witness(&mut self, bits: usize, witness: u32) -> Result<bool, &'static str> {
        if bits == 0 { return Ok(true); }
        self.absorb_field(witness)?;
        Ok(self.sample_bits(bits)? == 0)
    }
    pub fn grind(&mut self, bits: usize) -> Result<u32, &'static str> {
        if bits == 0 { return Ok(0); }
        for witness in 0..BABY_BEAR_MODULUS {
            let mut trial = self.clone(); trial.absorb_field(witness)?;
            if trial.sample_bits(bits)? == 0 { *self = trial; return Ok(witness); }
        }
        Err("no PoW witness")
    }
    pub fn state(&mut self) -> Result<[u8; 64], &'static str> { self.flush()?; Ok(self.state) }
    pub fn pending_items(&self) -> usize { self.pending.len() }
}

pub fn canonical(value: u32) -> Result<(), &'static str> { if value < BABY_BEAR_MODULUS { Ok(()) } else { Err("noncanonical field") } }
pub fn decode_field_array(payload: &[u8]) -> Result<Vec<u32>, &'static str> {
    if payload.len() < 4 { return Err("truncated count"); }
    let count = u32::from_be_bytes(payload[..4].try_into().map_err(|_| "count width")?) as usize;
    if payload.len() != 4usize.checked_add(count.checked_mul(4).ok_or("array size overflow")?).ok_or("array size overflow")? { return Err("array count/length mismatch or trailing bytes"); }
    payload[4..].chunks_exact(4).map(|word| { let value = u32::from_be_bytes(word.try_into().map_err(|_| "field width")?); canonical(value)?; Ok(value) }).collect()
}
fn validate_item(item_type: u8, payload: &[u8]) -> Result<(), &'static str> {
    match item_type { FIELD if payload.len() == 4 => canonical(u32::from_be_bytes(payload.try_into().map_err(|_| "scalar width")?)), FIELD => Err("scalar width"), COMMITMENT if payload.len() == 64 => Ok(()), COMMITMENT => Err("commitment width"), FIELD_ARRAY => decode_field_array(payload).map(|_| ()), _ => Err("unknown item type") }
}

fn abi_word(label: &[u8]) -> Result<[u8; 32], &'static str> { if label.len() > 32 { return Err("ABI word overflow"); } let mut out = [0u8; 32]; out[..label.len()].copy_from_slice(label); Ok(out) }
fn keccak(input: &[u8]) -> [u8; 32] { Keccak256::digest(input).into() }
pub fn statement_key(parameter: &[u8; 64], public_values: &[u32]) -> Result<[u8; 32], &'static str> {
    if public_values.len() != 64 { return Err("full-width statement required"); }
    let mut encoded = Vec::with_capacity(67 * 32); encoded.extend_from_slice(&abi_word(b"PQTC.V3.STATEMENT")?); encoded.extend_from_slice(parameter);
    for &value in public_values { canonical(value)?; encoded.extend_from_slice(&[0; 28]); encoded.extend_from_slice(&value.to_be_bytes()); }
    Ok(keccak(&encoded))
}
pub fn checkpoint_digest(key: &[u8; 32], payload: &[u8]) -> Result<[u8; 32], &'static str> {
    let mut encoded = Vec::with_capacity(96); encoded.extend_from_slice(&abi_word(b"PQTC.V3.CHECKPOINT")?); encoded.extend_from_slice(key); encoded.extend_from_slice(&keccak(payload)); Ok(keccak(&encoded))
}
pub fn proof_id(key: &[u8; 32], part_a: &[u8]) -> Result<[u8; 32], &'static str> {
    let mut encoded = Vec::with_capacity(96); encoded.extend_from_slice(&abi_word(b"PQTC.V3.PROOF")?); encoded.extend_from_slice(key); encoded.extend_from_slice(&keccak(part_a)); Ok(keccak(&encoded))
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FullDigestRecord { pub parameter: [u8; 64], pub statement_digests: [[u8; 64]; 4], pub global_digest: [u8; 32], pub core_proof_digest: [u8; 32] }
#[derive(Default)]
pub struct FullDigestStore { records: HashMap<[u8; 32], FullDigestRecord> }
fn validate_statement_digests(public_values: &[u32], digests: &[[u8; 64]; 4]) -> Result<(), &'static str> {
    if public_values.len() != 64 { return Err("full-width statement required"); }
    for (index, word) in digests.iter().flat_map(|digest| digest.chunks_exact(4)).enumerate() {
        let value = u32::from_be_bytes(word.try_into().map_err(|_| "statement field width")?);
        canonical(value)?;
        if value != public_values[index] { return Err("statement digest/public field mismatch"); }
    }
    Ok(())
}
impl FullDigestStore {
    pub fn put(&mut self, public_values: &[u32], record: FullDigestRecord) -> Result<[u8; 32], &'static str> {
        validate_statement_digests(public_values, &record.statement_digests)?;
        let key = statement_key(&record.parameter, public_values)?; self.records.insert(key, record); Ok(key)
    }
    pub fn get_checked(&self, key: &[u8; 32], public_values: &[u32], supplied: &FullDigestRecord) -> Result<&FullDigestRecord, &'static str> {
        let stored = self.records.get(key).ok_or("unknown lookup key")?;
        if &statement_key(&stored.parameter, public_values)? != key { return Err("stored full statement/key mismatch"); }
        validate_statement_digests(public_values, &stored.statement_digests)?;
        if stored != supplied { return Err("full-width digest mismatch"); }
        Ok(stored)
    }
}
